unist-util-map types Node and MapFunction stopped to be exported between v3 and v3.1.1

### Initial checklist

- [X] I read the [support docs](https://github.com/syntax-tree/.github/blob/main/support.md)
- [X] I read the [contributing guide](https://github.com/syntax-tree/.github/blob/main/contributing.md)
- [X] I agree to follow the [code of conduct](https://github.com/syntax-tree/.github/blob/main/code-of-conduct.md)
- [X] I searched [issues](https://github.com/search?q=user%3Asyntax-tree&type=Issues) and couldn’t find anything (or linked relevant results below)

### Affected packages and versions

unist-util-map 3.1.1

### Link to runnable example

https://stackblitz.com/edit/typescript-gjwbvb?file=index.ts

### Steps to reproduce

Start a new sandbox and install the latest `unist-util-map` v3.1.1:
https://stackblitz.com/edit/typescript-gjwbvb?file=index.ts
notice the `MapFunction` and `Node` (and `Parent`) types are not exported any more

but they used to be exported on v3:
https://stackblitz.com/edit/typescript-qh1hrn?file=index.ts

which means, we introduced a **breaking semver change** during a minor bump!

Now, other consumers such as `remark-directive-rehype` are importing the `unist-util-map` with a caret semver, which causes them to break on fresh installs ([stale lockfiles](https://github.com/IGassmann/remark-directive-rehype/blob/5dec9b98dcf8e071d2bd5b776f96123f459492f9/yarn.lock#L472) still contain working v3 references, so everything seems fine for those who have stale lockfiles and their package manager uses exact stale version v3)
https://github.com/IGassmann/remark-directive-rehype/blob/5dec9b98dcf8e071d2bd5b776f96123f459492f9/package.json#L34


### Expected behavior

Let's restore the old type exports, specifically:

```
export type Node = import('unist').Node
export type Parent = import('unist').Parent
export type MapFunction = (node: Node, index?: number, parent?: Parent) => Node
```

If we really want to remove the types, then issue a patch with types restored and bump the major version, publishing this program with types removed.

Generally, to prevent mishaps like this, it's best to migrate the source to `.ts` source files and let the TypeScript generate the types automatically, using a bundler. The JSDoc can be used for examples etc but not as a source of type definitions. That's the setup I use myself, I use `esbuild` to bundle ESM (and sometimes optionally IIFE too) and `rollup` with `rollup-plugin-dts` to bundle up the type definition files. The JSDoc-based types from `.js` sources are a liability.

### Actual behavior

type exports are missing currently

### Affected runtime and version

n/a

### Affected package manager and version

tested on both yarn and npm

### Affected OS and version

n/a

### Build and bundle tools

Other (please specify in steps to reproduce)

OK, I checked, it works in TS ^4.5, it does not work in 4.4 and lower. The root problem then is:

```js
index.js:6:21 - error TS1069: Unexpected token. A type parameter name was expected without curly braces.

6  * @template {Node} [Tree=Node]
```

The last release of TS 4.4 was on 2021-12-10. Do we think it’s fine to break that? (I am open to it)

--- a/index.js
+++ b/index.js
@@ -1,11 +1,12 @@
 /**
  * @typedef {import('unist').Node} Node
- * @typedef {import('unist').Parent} Parent
  */
 
 /**
- * @template {Node} [T = Node]
- * @typedef {import('./complex-types.js').MapFunction<T>} MapFunction<T>
+ * @template {Node} [Tree=Node]
+ * @typedef {import('./complex-types.js').MapFunction<Tree>} MapFunction
+ *   Function called with a node, its index, and its parent to produce a new
+ *   node.
  */
 
 /**

--- a/index.js
+++ b/index.js
@@ -1,5 +1,11 @@
 /**
  * @typedef {import('unist').Node} Node
+ * @typedef {import('unist').Parent} Parent
+ */
+
+/**
+ * @template {Node} [T = Node]
+ * @typedef {import('./complex-types.js').MapFunction<T>} MapFunction<T>
  */
 
 /**
@@ -9,7 +15,7 @@
  *   Type of input tree.
  * @param {Tree} tree
  *   Tree to map.
- * @param {import('./complex-types').MapFunction<Tree>} mapFunction
+ * @param {MapFunction<Tree>} mapFunction
  *   Function called with a node, its index, and its parent to produce a new
  *   node.
  * @returns {Tree}

--- a/readme.md
+++ b/readme.md
@@ -133,8 +133,7 @@ if the original node has children, those are mapped instead.
 ## Types
 
 This package is fully typed with [TypeScript][].
-It exports a type `MapFunction<Tree extends Node = Node>` from
-`unist-util-map/complex-types.d.ts` to properly type `MapFunction`s.
+It exports a type `MapFunction<Tree extends Node = Node>` to properly type `MapFunction`s.
 
 ## Compatibility
 

--- a/test.js
+++ b/test.js
@@ -46,10 +46,13 @@ test('unist-util-map', function (t) {
     'should work when passing an empty object'
   )
 
+  /** @type {Root} */
+  const tree = u('root', [u('node', [u('leaf', '1')]), u('leaf', '2')])
+
   t.deepEqual(
-    map(u('root', [u('node', [u('leaf', '1')]), u('leaf', '2')]), asIs),
-    u('root', [u('node', [u('leaf', '1')]), u('leaf', '2')]),
-    'should keep node as-is'
+    map(tree, asIs),
+    tree,
+    'should support an explicitly typed `MapFunction`'
   )
 
   t.end()
@@ -77,7 +80,7 @@ test('unist-util-map', function (t) {
   }
 
   /**
-   * @type {import('./index.js').MapFunction<AnyNode>}
+   * @type {import('./index.js').MapFunction<Root>}
    */
   function asIs(node) {
     return node

--- a/test.js
+++ b/test.js
@@ -46,6 +46,12 @@ test('unist-util-map', function (t) {
     'should work when passing an empty object'
   )
 
+  t.deepEqual(
+    map(u('root', [u('node', [u('leaf', '1')]), u('leaf', '2')]), asIs),
+    u('root', [u('node', [u('leaf', '1')]), u('leaf', '2')]),
+    'should keep node as-is'
+  )
+
   t.end()
 
   /**
@@ -68,5 +74,12 @@ test('unist-util-map', function (t) {
 
   function addValue() {
     return {value: 'test'}
+  }
+
+  /**
+   * @type {import('./index.js').MapFunction<AnyNode>}
+   */
+  function asIs(node) {
+    return node
   }
 })

